### DISCUSSION TITLE
Time delay before reporting no motion with radar modules

### DISCUSSION THREAD

**Discussion by dulfer** [ESPresense/ESPresense#462] [Ideas]
I have this feature request of which I am unsure if this is already built in (but I am incapable of finding the setting), so thought to open a discussion first. Also to poll whether someone else would appreciate this feature.

My ESPresense 'modules' all have RCWL-0516 Doppler Radar Motion Sensor Module connected that registers motion in that room. In contrary to most PIR modules, the RCWL-0516 itself does not have a time delay option to delay report after no motion is detected and resets after a second or so. This results in the state flapping between no-motion and motion detected whenever the room is occupied. 

I currently debounce the event in code, but since Espresense reports the status to the MQTT server it would be nice to add a setting to delay reporting the no motion detected for _x_ seconds (i.e. 5 seconds). If this window expires, it sends sends the state to MQTT broker. But if motion is detected within this window, the module does not send the _no motion detected_ state. 

Is such a setting already present? 
If not, would you appreciate if I create an Issue/FeatureRequest for it?

**Comment by DTTerastar** [comment]
I'm not opposed to a little debounce.  Instead of an issue just make a PR since you already have some code, right? :wink:

**Reply by dulfer** [reply]
cpp is not my weapon of choice, but I guess I should take this as a challenge 😉

**Reply by dulfer** [reply]
Small update to let you know I am working on it.
Still need to do some work on it before it is ready for a PR.

![image](https://user-images.githubusercontent.com/30404904/166483622-a74147c6-324c-4f96-839e-7576d84b46d4.png)

The graph below shows that the debouncing is working (peaks=debouncing for 5000ms) and reduces the number of false 'NO MOTION' events. Long gaps are real motionless periods.

![image](https://user-images.githubusercontent.com/30404904/166481846-0902255e-eb58-4e5b-a894-830f24c34614.png)

Would it make sense to keep publishing the debouncing state to an MQTT topic?
Currently implemented to allow for easy checking whether my code works and initially intended to take it out once it was ready for a PR. 

Next on the list is adjusting the debounce period from the Home Assistant device interface.

**Comment by dulfer** [comment]
@DTTerastar I have been running my modifications on my ESPPresence nodes for a week now. No issues so far.
I don't want to do a PR yet, as I am not satisfied with how I now pass the debounce delay parameter to the MotionSensors loop. 

Could you have a look at https://github.com/dulfer/ESPresense and advise if/how I can do this without passing the radarDebounceMs param in r612 (main.cpp)? I lack experience with c and could not figure out how to read the value directly from the config.

**Comment by DTTerastar** [comment]
For anyone looking this is in the beta [v2.2.0b0](https://github.com/ESPresense/ESPresense/releases/tag/v2.2.0b0) is it working correctly for everyone?

**Reply by dulfer** [reply]
Flashed all my devices with this release. Debounce works as expected and results in far better reports of actual room occupancy. 

After a day or so all of them were offline in HA, although the device's webservers were still accessible. A restart brought them back online. Haven't had this problem since, running without problems for almost a week now. Might have had another origin. 
Someone else recognises this behavior?
